test/inductor/test_torchinductor.py

```diff
@@ -243,6 +244,43 @@ def run(*ex, **kwargs):
                 if exact_dtype:
                     assert correct_val.dtype == actual_val.dtype
 
+    if check_gradient:
+        def gather_leaf_tensors(args, kwargs):
+            leaf_tensors = []
+            args, args_spec = tree_flatten(args)
+            kwargs, kwargs_spec = tree_flatten(kwargs)
+            args = args + kwargs
+            for arg in args:
+                if not isinstance(arg, torch.Tensor):
+                    continue
+                if arg.requires_grad:
+                    leaf_tensors.append(arg)
+            return leaf_tensors
+
+        def compute_grads(args, kwrags, results):
+            flat_results, _ = tree_flatten(results)
+            flat_diff_results = [r for r in flat_results if r.requires_grad]
+            assert len(flat_diff_results) > 0
+
+            grads = [torch.ones(r.shape, device=r.device, dtype=r.dtype) for r in flat_diff_results]
+
+            leaf_tensors = gather_leaf_tensors(ref_inputs, ref_kwargs)
+            assert len(leaf_tensors) > 0
+            return torch.autograd.grad(flat_diff_results, leaf_tensors,
+                                    grads, allow_unused=True, retain_graph=True)
+
+        correct_grad = compute_grads(ref_inputs, ref_kwargs, correct)
```

Comment:
should we be using gradcheck as @albanD suggested?

Reply:
IIUC, autograd.gradcheck computes gradients via small finite differences. It's useful for checking gradient's correctness when a reference implementation is not available. 

For inductor, we can use eager's result as the reference.

Reply:
The benefit of using gradcheck is that you don't have to write all the code to get proper grad outputs and do the comparison by hand.

Reply:
As discussed offline with Alban, I will stick with current approach, coz gradcheck doesn't work well with low precision floats.

Reply:
Good point about fp32 (not really low precision :-) )